### PR TITLE
Adds Tor support to cln-plugin

### LINKED ISSUE #92: Unable to connect to watchtowers through TOR
Hi there. I have 2 watchtowers running, one local and one remote. 

Unfortunately I must open a port and connect direct to the remote tower. This works as expected. I have TOR setup and get a TOR address when starting the watchtower. However local and remote instances of the watchtower-client are unable to connect.

lightning-cli registertower 021366c98eabdd5ecca68ceb4ebfcda58f24b1e6069e2c168c7c246b67e7827baa@7lqfpop2kbfplmg7vuuei5npfgyxmoh3du6zjlcgtu55ybp73hj2k2id.onion:2121
"Cannot connect to the tower. Connection refused"

TEOSD:
[teosd] Loading configuration from file
778Z INFO  [teosd] tower_id: 021366c98eabdd5ecca68ceb4ebfcda58f24b1e6069e2c168c7c246b67e7827baa
780Z INFO  [teosd] Last known block: 00000000000000000008708de7105faece0fbed942b0f601b971d87c62b4992a
780Z INFO  [teosd] Bootstrapping from backed up data
781Z INFO  [teosd] Bootstrap completed. Turning on interfaces
782Z INFO  [teosd] Starting up hidden tor service
782Z INFO  [teosd] Tower ready
784Z INFO  [teos::api::tor] onion service: 7lqfpop2kbfplmg7vuuei5npfgyxmoh3du6zjlcgtu55ybp73hj2k2id.onion:2121

Any pointers on how I can get TOR working? I already use tor on both local and remote machines for bitcoind and c-lighting which are working as expected. I can connect to bitcoind and lightning clients through tor but the watchtowers do not want to connect.

### PR DESCRIPTION
**Pull request by sr-gi**
Check if the tower address is an onion address and proxies it through Tor if so. 

This assumes Tor is running on the backend, which is the same assumption used by CoreLN AFAICT https://lightning.readthedocs.io/TOR.html

Close #92 #93

### CODE REVIEW

**Comment by sr-gi** [comment]
@CMDRZOD feel free to give this a go, it should fix your issue with connecting to the tower via `Tor`

**Comment by sr-gi** [comment]
Also tagging you @tee8z since you wrote the original piece of code. Can you give it a try?

**Comment by tee8z** [comment]
I gave this PR branch a test, seems I am getting "connection refused" when I try to connect from the plugin but works with tor-sock from command line. Attached what I saw when testing

tower-logs
torsock-works png
lightning-cli
plugin-logs

**Comment by sr-gi** [comment]
@tee8z actually got this same error but fixed it (locally at least) before pushing the last rebase 🤔

Are you testing this with both the client and the tower on the same host? What does your setup look like?

File: watchtower-plugin/src/net/http.rs
Comment (meryacine) [inline watchtower-plugin/src/net/http.rs:123]: Can't we ask cln to give us the SOCKS proxy host and port instead?
Or maybe provide them as options with `127.0.0.1:9050` as the default.

File: watchtower-plugin/src/net/http.rs
Comment (meryacine) [inline watchtower-plugin/src/net/http.rs:122]: This will fail if the tower's onion service is running on 443 (https) or 80 (http) and the port is not specified.
we can do something along the lines of:
```rust
let url = reqwest::Url::from(endpoint);
if url.host().unwrap().to_string().ends_with(".onion") {
    // some stuff
}
```

**Review by meryacine — CHANGES_REQUESTED** [review]
Some comments, but LGTM otherwise.
tACK
image

**Comment by meryacine** [comment]
@tee8z Are you sure you recompiled the cln plugin before running `lighningd`?
You might be running an older version of cln plugin.

File: watchtower-plugin/src/net/http.rs
Comment (sr-gi) [reply watchtower-plugin/src/net/http.rs:123]: I'm working on it with the second approach, but it becomes a bit more complex.

Thought about asking cln, but that'd mean that if cln is using `Tor` the plugin always will, and if it is not, the plugin would never. I think we may want to give the user the option to pick any combination.

File: watchtower-plugin/src/net/http.rs
Comment (sr-gi) [reply watchtower-plugin/src/net/http.rs]: Fixed in [dcbcee3](https://github.com/talaia-labs/rust-teos/pull/96/commits/dcbcee385a45aaf36abf1d823fa6e79ae941325b)

File: watchtower-plugin/src/net/http.rs
Comment (sr-gi) [reply watchtower-plugin/src/net/http.rs]: Fixed in [dcbcee3](https://github.com/talaia-labs/rust-teos/pull/96/commits/dcbcee385a45aaf36abf1d823fa6e79ae941325b). Not completely convinced it is the most elegant way though.

**Comment by sr-gi** [comment]
Looks like there are two tests failing, given we used to flag malformed urls as `Unexpected` errors instead of `Connection`. Will need to either repurpose or remove those tests.

**Comment by meryacine** [comment]
> Looks like there are two tests failing, given we used to flag malformed urls as Unexpected errors instead of Connection. Will need to either repurpose or remove those tests.

Dunno what cases now trigger the `Unexpected` path. I think we can remove them if no testable path possible.

**Review by meryacine — APPROVED** [review]
LGTM.
One thing I did think of though, is to have `127.0.0.1:9050` be the default proxy (we won't have `None`s in this case). But actually both techniques are perfectly fine.

**Comment by sr-gi** [comment]
> LGTM. One thing I did think of though, is to have `127.0.0.1:9050` be the default proxy (we won't have `None`s in this case). But actually both techniques are perfectly fine.

The default should be to have no proxy at all, shouldn't it? Otherwise, you'd be trying to redirect your traffic to an inexistent proxy.

**Comment by sr-gi** [comment]
This needs rebasing all commits into a single one. Will wait for @tee8z review before doing so and potentially merging.

**Comment by tee8z** [comment]
I'm sure this is something with my setup, but I'm still struggling to get the plugin to work with tor, new error message though! 
From my testing I can see that:
* I have tor running and able to call through it to an external address via socks5
* I have lightningd running and it's registering with tor (showing a tor address)
* This is all running over signet (shouldn't change anything?)
* The tower is running and showing a tor address to connect with

I'm going to keep trying to get this over the coming days, but don't let me hold up merging this in as it's probably specific to my environment.

ksnip_20220824-100955
lightningd-running
conf-setup
tower-setup
socks5-working

**Comment by tee8z** [comment]
Changing the setting to `watchtower-proxy` and clearing any local data for the tower made it work! Thank you for the helpful direction, otherwise code LGTM!
it-worked

**Review by tee8z — APPROVED** [review]
LGTM

**Comment by sr-gi** [comment]
Nice! I think I'm going to improve this a bit and use `cln` native options instead of adding a custom one. I'm waiting to confirm what's the logic behind `cln's` `proxy` and `always-use-proxy` so I can adapt this accordingly.

**Comment by sr-gi** [comment]
> Nice! I think I'm going to improve this a bit and use `cln` native options instead of adding a custom one. I'm waiting to confirm what's the logic behind `cln's` `proxy` and `always-use-proxy` so I can adapt this accordingly.

Ok, nvm, this will need to be added in a follow-up given the native `cln` options cannot be accessed from the `cln-plugin` rust crate atm AFAICT.

I'm documenting the improvements that we can land after the next `cln-plugin` in release https://github.com/talaia-labs/rust-teos/discussions/102

**Pull request merged**
